### PR TITLE
Show style applied on tag (p, span, div) and private selectors as parent rules instead of hiding them

### PR DESCRIPTION
**Pull request by Dobby85**
### Current behavior

The current behavior describe below can be tested on this JSFiddle : https://jsfiddle.net/728urdh3/

#### Context

I add a custom component (`sectionBlock`) that have styles by the `section` tag name and the private selector `ed-layout-div`.

```css
section {
  width: 940px;
  min-height: 50px;
  border: 1px dashed black;
}

.ed-layout-div {
  margin-top: 20px;
  margin-bottom: 30px;          
}
```

#### Step to reproduce

When you drop this component in the canvas and check it's style.

- You **do not see** `width` and `min-height` properties set on the section.
- You **see** the `margin-top` and the `margin-bottom` apply on the hidden selectors.

Screenshot 2024-05-17 at 08 25 43

Now when you add a class to the component :

- You **do not see anymore** the `margin-top` and the `margin-bottom` apply on the hidden selectors.

Screenshot 2024-05-17 at 08 27 20

#### Context

The context is the same than the previous one

#### Step to reproduce

When you drop this component in the canvas and check it's style.

- You **see** `width` and `min-height` properties set on the section as parent rules
- You **see** `margin-top` and `margin-bottom` apply on the hidden selectors as parent rules

Screenshot 2024-05-17 at 08 28 16

Now when you add a class to the component :

- You still **see** `width` and `min-height` properties set on the section as parent rules
- You still **see** `margin-top` and `margin-bottom` apply on the hidden selectors as parent rules

Screenshot 2024-05-17 at 08 28 31

### Why do I do that?

I think it's more comfortable to always see the style that is apply on the component to better understand why it is like it is. You still can override style but you always know which rule are applied to the component.

All tests passed and I add 3 new tests to test style apply on tag name and return style with multiple classes.

Do not hesitate if you need more information or if I have to update/fix something !

Have a great day!

### CODE REVIEW

File: src/style_manager/index.ts [hunk 1]
```diff
@@ -597,27 +597,65 @@ export default class StyleManager extends ItemManagerModule<
...
+          const rSels = rule.getSelectors().map(s => s.getFullName());
+
+          // rSels is equal to 0 when rule selectors contain tagName like : p {}, .logo path {}, ul li {}
+          if (rSels.length === 0) {
+            return false;
+          }
+
+          return rSels.every(rSel => values.indexOf(rSel) >= 0);
+        });
+      };
```
Comment (artf) [inline src/style_manager/index.ts:616]: What is the point of this change and why did you remove `!values.length ? []`? That was important to avoid performance issues. 
We don't want to loop over all the rules if `values` is empty

File: src/style_manager/index.ts [hunk 2]
```diff
@@ -597,27 +597,65 @@ export default class StyleManager extends ItemManagerModule<
...
+
+      const rulesByTagName = (tagName: string) => {
+        return cssC.getRules().filter(rule => {
+          const ruleTags = rule
+            .selectorsToString()
+            .split(' ')
+            .filter(item => {
+              return item.startsWith('.') === false && item.startsWith('#') === false;
```
Comment (artf) [inline src/style_manager/index.ts:624]: why not to simply check `rule.selectorsToString() === tagName`?
The logic here will take wrong rules, eg. `.something DIV .something-else`

File: src/style_manager/index.ts [hunk 3]
```diff
@@ -597,27 +597,65 @@ export default class StyleManager extends ItemManagerModule<
...
       // Componente related rule
       if (cmp) {
         cmpRules = cssC.getRules(`#${cmp.getId()}`);
+        tagNameRules = rulesByTagName(cmp.get('tagName'));
         otherRules = sel ? rulesBySelectors(sel.getSelectors().getFullName(optsSel)) : [];
-        rules = otherRules.concat(cmpRules);
+        rules = otherRules.concat(tagNameRules).concat(cmpRules);
       } else {
         cmpRules = sel ? cssC.getRules(`#${sel.getId()}`) : [];
+        tagNameRules = rulesByTagName(sel?.get('tagName') || '');
```
Comment (artf) [inline src/style_manager/index.ts:645]: Keep in mind to avoid useless loops when it's possible

File: src/style_manager/index.ts [hunk 4]
```diff
@@ -597,27 +597,65 @@ export default class StyleManager extends ItemManagerModule<
...
+        // Get rules set on invisible selectors like private one
+        const allCmpClasses = sel?.getSelectors().getFullName(optsSel) || [];
+        const invisibleSel = allCmpClasses.filter(
+          (item: string) =>
+            target
+              .getSelectors()
+              .getFullName(optsSel)
+              .findIndex(sel => sel === item) === -1
+        );
+        invisibleRules = rulesBySelectors(invisibleSel);
```
Comment (artf) [inline src/style_manager/index.ts:657]: `rulesBySelectors` is already executed above, it would make sense to combine CssRule selectors + Component ones and pass them to `rulesBySelectors` together

File: test/specs/style_manager/index.ts [hunk 5]
```diff
@@ -195,6 +195,44 @@ describe('StyleManager', () => {
...
+        const [rule1, rule2, rule3] = cssc.addRules(`
+          .cls { color: red; }
+          div { color: yellow; }
+          #id-test { color: blue; }
+        `);
+        em.setSelected(cmp);
+        obj.__upSel();
+        expect(obj.getSelected()).toBe(rule1);
+        expect(obj.getSelectedParents()).toEqual([rule3, rule2]);
+      });
```
Comment (artf) [inline test/specs/style_manager/index.ts:234]: Thanks 🙇‍♂️ it's good to cover with tests this kind of logic

**Review by artf — CHANGES_REQUESTED** [review]
Thank you @Dobby85 that's a nice idea and overall a good PR.
I'd just ask you to review my comments and keep performance in mind here to avoid falling into [similar issues](https://github.com/GrapesJS/grapesjs/issues/5525).

File: src/style_manager/index.ts (hunk 1 above)
Comment (Dobby85) [reply src/style_manager/index.ts]: My bad, I don't why I remove it, I just add it again!

File: src/style_manager/index.ts (hunk 2 above)
Comment (Dobby85) [reply src/style_manager/index.ts]: I cannot just check `rule.selectorsToString() === tagName` because `tagName` only include the tag name like `div`, `section`, `li`... and not all the path like `rule.selectorsToString()` can have (`.something .something-else div`).

In fact, it's a problem for this case: `.something DIV .something-else`. So I add a check to be sure the tagName is the last element in the `rule.selectorsToString()`.

I know this function is not perfect, it still contains this caveat:

```html
<section class="cls1">
  <div>Text</div>
</section>
```
```css
.cls1 div {
  padding: 10px;
}
```

In the previous case,`.cls1 div` will be return as a rule for both div because I don't how to check a component is a child or not of a component having `.cls1`. Or if we can check that I think we will add a lot of loops...

I don't know what is the best things to do:

1 - Know the caveat and let it like this
2 - Try to fix it and add a lot of checks and loops

In every case, I fix the following case `.something DIV .something-else` and add a test on it :)

File: src/style_manager/index.ts (hunk 4 above)
Comment (Dobby85) [reply src/style_manager/index.ts]: Of course, I update code to call `rulesBySelectors` only once!

File: test/specs/style_manager/index.ts (hunk 5 above)
Comment (Dobby85) [reply test/specs/style_manager/index.ts:248]: It's always better when it's tested 😉

**Comment by Dobby85** [comment]
@artf I don't know who is supposed to resolve conversation. I fix everything I could based on your comment, I let you review changes and resolve conversation if it sounds good to you ☺️

File: src/style_manager/index.ts (hunk 2 above)
Comment (artf) [reply src/style_manager/index.ts]: I'd suggest to keep it simple for now, only selectors with one and single tagName are accepted, eg. `div` is ok, `.cls1 div` is not.  So let's just do `rule.selectorsToString() === tagName`.

You're not considering a lot of CSS selectors, like in your case `.cls2 div` will be taken as valid, also `.cls2 div div`... it's better to skip those rules instead of showing the wrong ones. 
We won't be able to be precise with the current approach anyway.

File: src/style_manager/index.ts (hunk 3 above)
Comment (artf) [reply src/style_manager/index.ts:645]: Let's place a check 
```js
const rulesByTagName = (tagName: string) => {
        return !tagName ? [] : cssC.getRules().filter(rule => {
        ...
```

Comment (Dobby85) [reply src/style_manager/index.ts:632]: I add the check 👍

**Comment by Dobby85** [comment]
All your comments should be resolve @artf. Tell me if it's not! Else I think everything is ready for merge 😁

**Review by artf — APPROVED** [review]

**Comment by artf** [comment]
Nice, thanks @Dobby85

File: src/style_manager/index.ts (hunk 2 above)
Comment (chen-can) [reply src/style_manager/index.ts]: The 0.17 version is supported.

**Comment by chen-can** [comment]
> 您的所有意见都应该得到解决@artf。如果不是的话请告诉我！不然我想一切都已经准备好进行合并了😁

 <style>
        .mw-package-price,
        .mw-package-price span.price {
          color: red;
        }

        .mw-package-price span {
          color: #4d4d4d;
        }
      </style>
      <p class="mw-package-price">
        <span class="regular-price">$98.00</span>
        <span class="price" id="prod-1-price">$49.00</span>
      </p>
      fail

**Comment by chen-can** [comment]
Version 0.17.29 works and matches correctly, but the newer version does not. I need to use the newer version because grapesjs-style-bg requires a better version
